Parse/Internal/LocalDataStore/OfflineStore/PFOfflineStore.m

```diff
@@ -499,52 +495,59 @@ - (BFTask *)findAsyncForQueryState:(PFQueryState *)queryState
                                      PFOfflineStoreKeyOfUUID, PFOfflineStoreTableOfObjects,
                                      PFOfflineStoreTableOfDependencies, PFOfflineStoreKeyOfUUID,
                                      PFOfflineStoreKeyOfUUID, PFOfflineStoreKeyOfClassName,
                                      PFOfflineStoreKeyOfKey, isDeletingEventuallyQuery];
+            queryArguments = @[ queryState.parseClassName, uuid ];
 
-            return [database executeQueryAsync:queryString
-                          withArgumentsInArray:@[ queryState.parseClassName, uuid ]];
+            return nil;
         }];
     }
 
     @weakify(self);
     return [[queryTask continueWithSuccessBlock:^id(BFTask *task) {
-        @strongify(self);
-        PFSQLiteDatabaseResult *result = task.result;
-
-        PFConstraintMatcherBlock matcherBlock = [self.offlineQueryLogic createMatcherForQueryState:queryState
-                                                                                              user:user];
-
-        BFTask *checkAllObjectsTask = [BFTask taskWithResult:nil];
-        while ([result next]) {
-            NSString *uuid = [result stringForColumnIndex:0];
-            __block PFObject *object = nil;
-
-            checkAllObjectsTask = [[[[checkAllObjectsTask continueWithSuccessBlock:^id(BFTask *task) {
-                return [self _getPointerAsyncWithUUID:uuid database:database];
-            }] continueWithSuccessBlock:^id(BFTask *task) {
-                object = task.result;
-                return [self fetchObjectLocallyAsync:object database:database];
-            }] continueWithSuccessBlock:^id(BFTask *task) {
-                if (!object.dataAvailable) {
-                    return @NO;
-                }
-                return matcherBlock(object, database);
-            }] continueWithSuccessBlock:^id(BFTask *task) {
-                if ([task.result boolValue]) {
-                    [mutableResults addObject:object];
-                }
-                return nil;
-            }];
-        }
-        [result close];
-
-        return checkAllObjectsTask;
-    }] continueWithSuccessBlock:^id(BFTask *task) {
+        return [[database executeQueryAsync:queryString withArgumentsInArray:queryArguments block:^id(PFSQLiteDatabaseResult *result) {
+            NSMutableArray *uuids = [NSMutableArray array];
+            while ([result next]) {
+                NSString *uuid = [result stringForColumnIndex:0];
+                [uuids addObject:uuid];
+            }
+            return uuids;
+        }] continueWithSuccessBlock:^id(BFTask PF_GENERIC(NSArray *)*task) {
+            @strongify(self);
+            PFConstraintMatcherBlock matcherBlock = [self.offlineQueryLogic createMatcherForQueryState:queryState user:user];
+
+            BFTask *checkAllTask = [BFTask taskWithResult:nil];
+            NSArray *uuidBatches = [PFInternalUtils arrayBySplittingArray:task.result withMaximumComponentsPerSegment:25];
```

Comment:
We split to make sure we do not allocate a lot of memory. Will change the limit to 64 to keep it at power of 2.